Wait for CEC and CCEC resources before restoring endpoints.

jrajahalme opened the pull request:
Just like we are currently waiting for policy CRDs before generating endpoints for the first time after restoring them during an agent restart, it is also beneficial to wait for CEC and CCEC resources as well. This is mainly due to possible policy redirections to CEC/CCEC listeners, but also for the possibility of L7 LB service annotations, as well as east/west Ingress and Gateway API, which are implemented via CEC/CCEC resources.

In order to achieve this we first break up potentially blocking parts of daemon lifecycle start hook to run in a goroutine, so that the hive may run concurrently with the daemon bootstrap. This is necessary as start hooks are executed in the order in which they are appended to the hive lifecycle, and the next start hook can only run when the previous one has returned.

startDeamon() blocks waiting for a subset of k8s resources to have been synchronized, and as we add CEC and CCEC resources to that subset, we want to be able to use the hive lifecycle to start running the resource watchers for these resources. This becomes possible by making sure daemon start hooks do not block for this purpose. Similarly, any start hook that needs to await for the daemon promise needs to perform that Await call in a goroutine. This allows the hive to run all the needed start hooks while startDaemon() is waiting for the resources to have been synchronized.

This change also allows reverting the change in k8s policy watchers introduced in https://github.com/cilium/cilium/pull/32028, as now the hive is not stalled by the daemon waiting for the k8s resources to synchronize.

During testing there were spurious error and warning logs from the vendored k8s code when Cilium resources were requested before the corresponding CRDs had been registered. This is resolved by providing a new CRDSync promise to the hive, and then awaiting for that promise before requesting any Cilium resources. This eliminates the error and warning logs seen for CiliumNode, CiliumEnvoyConfig, and CIliumClusterwideEnvoyConfig resources. The CRDSync promise is optional so that it can be ignored for other than the agent hive.

jrajahalme committed:
k8s: Wait for CEC/CCEC sync on bootstrap

Include CEC and CCEC into the set of k8s resources that must be
synchronized before endpoints are regenerated. This reduces endpoint
regenerations during restart and reduces L7 policy churn.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

gandro commented on pkg/policy/k8s/cell.go:
```diff
@@ -65,69 +66,63 @@ type PolicyWatcherParams struct {
+			policyManager, err := p.PolicyManager.Await(startCtx)
+			if err != nil {
+				return err
+			}
```
We can't revert this commit as is. `registerResourceWithSyncFn` (called in `w.watchResources(ctx)`) **needs** to be called before `InitK8sSubsystem`. If we wait for the Daemon promise here, then `watchResources` is called after `InitK8sSubsystem` and thus we're not waiting for policy ingestion to initiate endpoint regeneration.

I think we need https://github.com/cilium/cilium/pull/32847 before we can revert this commit

jrajahalme replied:
Good call! I already had done the `registerResourceWithSyncFn` for CiliumEnvoyConfig (3rd commit) in the constructor instead of the start hook, just for this reason.

Just force pushed new 2nd commit that adds that change to the revert, please take a look!

gandro commented on pkg/ciliumenvoyconfig/cell.go:
```diff
@@ -87,6 +92,13 @@ func registerCECK8sReconciler(params reconcilerParams) {
 		},
 	})
 
+	reconciler.registerResourceWithSyncFn(context.Background(), k8sAPIGroupCiliumEnvoyConfigV2, func() bool {
```
Is it guaranteed that this is called before `InitK8sSubsystem`?

jrajahalme replied:
Yes, as all the constructors are called before any of the start hooks are run (if any constructor fails, the hive is not run at all and no start hooks are run). Changed network policy watcher to do the same.

gandro commented on daemon/cmd/daemon_main.go:
```diff
@@ -1719,16 +1719,44 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					if err := startDaemon(daemon, restoredEndpoints, cleaner, params); err != nil {
```
Are we sure there are no start hooks which implicitly relied on `startDaemon` (and thereby endpoint regeneration) to be completed? In particular ones that are relying on `option.DaemonConfig`, which I believe we have a few.

jrajahalme replied:
Not 100% sure, but it looks like it. No-one has access to the daemon before startDaemon completes as the promise is resolved... or at least that is how it should be :-)

gandro replied:
I don't think we can rely on type safety here unfortunately. For example, any cell that has access to the `EndpointManager` and `DaemonConfig` was previously implicitly waiting for endpoint regeneration to finish. Now, that start hooks executed _before_ endpoint regeneration, which means they might observe an empty endpoint manager.

One doesn't need to have a reference to `*Daemon` in order to observe changes in behavior with this commit

jrajahalme replied:
So maybe we need to add a promise for EndpointManager to fix this, or make the users of EndpointManager await for the restorer promise? Can you think of a systematical way we could find issues with this?

jrajahalme replied:
So this concern only applies to cells that are dependent on `DaemonConfig`, as it is now resolved earlier than the Daemon promise - this narrows down the concern quite a bit: go through dependents of `DaemonConfig` and add an await for the Daemon promise when needed?

Or is this dependency transitive, if one cell depends on `DaemonConfig` and a later cell then depends on `EndpointManager`, then that cell may see an empty endpoint manager?

jrajahalme replied:
There is only one dependency on`promise.Promise[*option.DaemonConfig]` (pkg/maps/nat), which is then providing `promise.Promise[NatMap4], promise.Promise[NatMap6]` for which there do not seem to be significant dependencies:
```
% git grep "Promise.*NatMap"      
pkg/datapath/fake/cells.go:             func() (promise.Promise[nat.NatMap4], promise.Promise[nat.NatMap6]) {
pkg/maps/nat/cell.go:   cell.Provide(func(lc cell.Lifecycle, cfgPromise promise.Promise[*option.DaemonConfig]) (promise.Promise[NatMap4], promise.Promise[NatMap6]) {
pkg/maps/nat/stats/stats.go:    NatMap4   promise.Promise[nat.NatMap4]
pkg/maps/nat/stats/stats.go:    NatMap6   promise.Promise[nat.NatMap6]
pkg/maps/nat/stats/stats_test.go:               cell.Provide(func() (promise.Promise[nat.NatMap4], promise.Promise[nat.NatMap6], Config, natMetrics) {
```

gandro replied:
Thanks for checking! Gives us much more confidence in the PR

tommyp1ckles replied:
> There is only one dependency onpromise.Promise[*option.DaemonConfig] (pkg/maps/nat), which is then providing promise.Promise[NatMap4], promise.Promise[NatMap6] for which there do not seem to be significant dependencies:

For context, this is a recent change, this code depends on a stable `option.Config.EnableNodePort` which can change during the newDaemon/startDaemon phases. 

However there are likely other modules that may have potentially racey behavior due to the volatile nature of DaemonConfig.

Here's a issue I created for the problem: https://github.com/cilium/cilium/pull/32558

tommyp1ckles replied:
To elaborate here, this might cause problems as startDaemon runs `finishKubeProxyReplacementInit` which, for example, can [disable options such as & related to](https://github.com/cilium/cilium/blob/fb7e01b4875cbb47b9c8cec8c4da3dfd96ba6490/daemon/cmd/kube_proxy_replacement.go#L349) `option.Config.EnableNodePort` (which currently must be in startDaemon because it waits for device detection to complete).

So the dependencies on `promise.Promise[*option.DeamonConfig]` (currently just pkg/maps/nat) depend on this promise resolving because they need to know that `option.Config`is finalized before proceeding: 

https://github.com/cilium/cilium/blob/fb7e01b4875cbb47b9c8cec8c4da3dfd96ba6490/pkg/maps/nat/cell.go#L39-L43

tommyp1ckles replied:
Since currently there is just that one dependency on the DaemonConfig promise, could we just resolve it at the same time as we do daemonPromise below?

I think that should solve any issues.

jrajahalme replied:
`finishKubeProxyReplacementInit()` is called from `newDaemon()`, before the DaemonConfig promise is resolved. AFAIK `option.Config` is not modified in `startDaemon()`.

gandro commented on pkg/ciliumenvoyconfig/cell.go:
```diff
@@ -87,6 +92,13 @@ func registerCECK8sReconciler(params reconcilerParams) {
+	reconciler.registerResourceWithSyncFn(context.Background(), k8sAPIGroupCiliumEnvoyConfigV2, func() bool {
+		return reconciler.cecSynced.Load()
+	})
+	reconciler.registerResourceWithSyncFn(context.Background(), k8sAPIGroupCiliumClusterwideEnvoyConfigV2, func() bool {
```
It's not ideal that we're using `context.Background()` here. I think this might leak go routines if this cell is ever included by the Cell-based tests in `test/controlplane`

Can we use a context that is cancelled once the cell is stopped (like the CiliumNetwork policy watcher does)?

jrajahalme replied:
Done :-)

gandro reviewed (commented):
The second commit looks fine to me now. I still feel a bit uneasy about the fact that `go startDaemon()` might cause lifecycle start hooks earlier than before.

jrajahalme committed:
daemon: Avoid blocking during hive start

Execute startDaemon() in a goroutine, which resolves daemon promise when
done. Change dependent start hooks to await for daemon promise in
goroutines so that the execution of start hooks is not stalled. This way
the hive may run concurrently with daemon start.

Move some early validation from startDaemon() to the start hook so that
we can fail out while we can simply return the error from the start
hook.

Move saving of daemon config to file also away from startDaemon, so that
we can resolve the DaemonConfig promise from the start hook itself. The
daemon promise will be resolved from the goroutine that runs
startDaemon(), or rejected if there was any error.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
Revert "policy/k8s: Fix bug where policy synchronization event was lost"

This reverts commit 3f9c28810b543bf95d8c169c73f670fc45a9726e.

In addition to the reject:

- registerResourceWithSyncFn() are moved from the start hook
  to the constructor so that they are called before
  daemon.InitK8sSubsystem().

- PolicyManager promise is resolved by newDaemonPromise() after Daemon
  has been constructed, but before the daemon waits for the k8s resources
  to have synced.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

gandro commented on daemon/cmd/deletion_queue.go:
```diff
@@ -26,26 +27,39 @@ var deletionQueueCell = cell.Group(
-	}
+		// hook context cancels when the start hooks have run, use context.Background()
+		// as we may be running after that.
+		d, err := dq.daemonPromise.Await(context.Background())
```
Please use the context provided by `Start()` here

jrajahalme replied:
I had CI break with the `ctx` from `Start()` here, and had to use `context.Background()` instead.

The context provided by `Start()` is cancelled after all the start hooks have executed, and the daemon promise is likely only resolved after that (that's what the comment says).

gandro replied:
Ah, I missed the comment sorry. We have lots of other code still using the start `ctx` for  `dq.daemonPromise.Await`. You need to fix those up as well then I guess?

joamaki replied:
This change here also seems to change the behavior. We're no longer handling the error from `processQueuedDeletes` and continuing with starting when earlier we stopped if it failed. This also runs now asynchronously with everything else rather than sequentially during start-up. Does this break something?

And yeah, as @gandro said, all `Await`s for daemon promise will now need to use a context that lives beyond `Start`.

jrajahalme replied:
Checked again, no start context use in daemon promise `Await()` any more.

gandro commented on daemon/cmd/daemon_main.go:
```diff
@@ -1905,28 +1934,28 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				daemon, err := daemonPromise.Await(context.Background())
```
Please use the context provided by `Start()` here

jrajahalme replied:
Could have been possible before this change (`context.Background()` was already used before), but now that the `daemonPromise` can resolve after all the start hooks have executed and the given context is cancelled, it is no longer possible to use that context for the `Await` here.

gandro reviewed (approved):
Unblocking the PR from a policy perspective (two suggestions remaining, but they are not fundamental). I'm assuming the foundations team will also take a look at the overall change here

jrajahalme committed:
daemon: Check that DaemonConfig is not changed after being published

Add sha256 sum to config and validate no change between DaemonConfig
promise and Daemon promise are resolved.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras reviewed (approved):
:astonished:

pippolo84 commented on pkg/k8s/synced/resources.go:
```diff
@@ -66,6 +69,11 @@ func (r *Resources) BlockWaitGroupToSyncResources(
+	if r.CacheStatus != nil && r.CacheStatus.Synchronized() {
+		log.WithField("kubernetesResource", resourceName).Errorf("BlockWaitGroupToSyncResources called after Caches have already synced")
+		return
+	}
```
Can't we avoid this test-specific check by initializing the Resources in `TestWaitForCacheSyncWithTimeout` with:

```go
r := &Resources{CacheStatus: make(k8s.CacheStatus)}
```

at pkg/k8s/synced/resources_test.go:93 ?

jrajahalme replied:
Yes! To be included in the next push!

pippolo84 reviewed (approved):
Just left a minor thing inline.

Regarding the approach, I cannot think about a better non-intrusive alternative for now, so there's no reason to block the PR from my side.

jrajahalme committed:
k8s: Move CastInformerEvent to informer

Move CastInformerEvent function to k8s/informer. Users of this function
already import k8s/informer, and this allows some of them to remove
import of 'k8s' itself.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
k8s/synced: Log an error if a resource is added to blocking set too late

Log an error so that CI will flake when a resource is added to the
blocking set after caches have already synchronized.

Moved provision of k8s.CacheStatus type to k8s/synced and provision it in
k8s/synced to avoid circular dependency.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
resources: Wait for CRDs before listing Cilium resources

Wait for CRD sync before trying to list Cilium resources. This eliminates
error and warning logs from vendored k8s code that happen when an unknown
resource is listed or watched.

CRDSync promise is optional. This avoids hanging any hives that do not
provide this promise.

Add a once-a-minute info log on the remaining CRDs not yet
registered. This may help debug cases when we are waiting for CRDs that
are not going to be registered at all.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

dylandreimerink commented on daemon/cmd/daemon_main.go:
```diff
@@ -1735,6 +1737,45 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
+		return nil
+	}
+
+	crdSyncResolver, crdSyncPromise := promise.New[resource.CRDSync]()
```
Adding null-able values into the dependency graph isn't best practice, if at all possible we like to avoid it to reduce the chances of a nil-deref if anyone uses this type in the future and doesn't realize it needs to be nil checked.

In this case I would suggest to always return a promise. But to `Reject` on it if we are never going to resolve it. Rejected promises return an error and don't block.

jrajahalme replied:
This promise is already optional, so we must `nil` check it anyway. Can reject and return the promise, however.

dylandreimerink commented on pkg/k8s/resource_ctors.go:
```diff
@@ -51,6 +52,19 @@ func (def Config) Flags(flags *pflag.FlagSet) {
+
+	Lifecycle      cell.Lifecycle
+	ClientSet      client.Clientset
+	CrdSyncPromise promise.Promise[resource.CRDSync] `optional:"true"`
```
Optional and nil-able are not the same.

The `optional` annotation instructs hive to not crash if `promise.Promise[resource.CRDSync]` isn't provided. It isn't needed here, even if the cell that provides it were to return `nil`. 

Actually, this would cause a silent error if the type is not provided instead of error-ing in CI and telling us about it.

jrajahalme replied:
IMO this needs to be optional, as it may not be provided in other than agent hives. Other hives also use the k8s package, and I did not want to go and find all the different users and provide a dummy promise in those.

jrajahalme replied:
So the background is that the k8s package is used by operator, and clustermech-apiserver, but it contains agent specific code for waiting for CRD sync, which is also used for this promise. Having the promise be optional makes it possible to those other binaries not have to wait for Agent's CRDs to be registered before observing a subset of those resources. We could make this non-optional, but then we'd have to make sure we provide a suitably different version on each different hive, implementations for which are not readily available.

I think we can alleviate the missing promise concern by not checking for `nil` in the daemon itself, so that it will crash if not provided.

jrajahalme replied:
Made `CRDSync` promise required for the agent hive, optional for other hives.

dylandreimerink replied:
> IMO this needs to be optional, as it may not be provided in other than agent hives.

Understandable, I had the causality the other way around. 

> Made CRDSync promise required for the agent hive, optional for other hives.

Thanks, that should catch most of the issues I was worried about.

dylandreimerink commented on daemon/cmd/daemon_main.go:
```diff
@@ -1704,7 +1704,15 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
+		OnStart: func(cell.HookContext) (err error) {
+			defer func() {
+				// Reject promises on error
+				if err != nil {
```
Does this still work if you shadow `err`? like in `d, restoredEndpoints, err := newDaemon(daemonCtx, cleaner, &params)`. 

As far as I know, the moment you shadow declare with `:=` the `err` in the defer and that below the new declaration are not the same anymore.

jrajahalme replied:
They'll be the same if returned, I think, i.e., return assigns to the return value `err`, and that is what we are inspecting here. Works for cases like `return nil`, etc.

dylandreimerink commented on daemon/cmd/daemon_main.go:
```diff
@@ -1714,16 +1722,44 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
 			}
-			daemonResolver.Resolve(daemon)
+
+			// NOTE: 'option.Config' must not be modified after this point
```
The `option.Config` currently can be modified during the entire runtime take `option.Config.Opts` for example. So this isn't entirely accurate.

jrajahalme replied:
Are there any other fields that are known to be runtime configurable? `Opts` seems to be the only one with a comment that indicates that:
```
	// Options changeable at runtime
	Opts *IntOptions
```

I'll make this ignored for the shasum.

jrajahalme replied:
Done.

dylandreimerink replied:
That was the only one I know for sure. Lets assert the rest is stable and see what errors shake loose

dylandreimerink commented on daemon/cmd/daemon_main.go:
```diff
@@ -1981,6 +1981,19 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
 	bootstrapStats.updateMetrics()
 	go d.launchHubble()
 
+	// Start controller to validate daemon config is unchanged
```
Not sure this is going to work, see earlier comment

jrajahalme replied:
Will fix. I did not inspect the controller status for errors yet, though. Do you happen to know of a code in the CI that could be used to surface controller errors?

jrajahalme replied:
Ginkgo suite checks for controller errors in Agent preflight checks on `ExpectCiliumReady()`, so there is some coverage, but none in the `cilium-cli` test suite.

jrajahalme committed:
daemon: Do not stall hive Stop if k8s caches are never synced

Make wait on k8s cache sync conditional on the daemon context not being
canceled.

Apply same logic in the goroutine waiting for endpoint restoration.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
option: Make ConfigPatchMutex a pointer

Make ConfigPatchMutex a pointer to be able to take a shallow copy for
summing the immutable parts of Config (everything else than the
IntOptions).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -1737,6 +1739,46 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
+	APIGroups *k8sSynced.APIGroups
+}
+
+func newSyncCRDsPromise(params syncCRDsPromiseParams) promise.Promise[resource.CRDSync] {
```
Consider putting this into `daemon/k8s` next to the other related code.

jrajahalme replied:
Done.

joamaki commented on test/controlplane/suite/agent.go:
```diff
@@ -74,6 +75,11 @@ func (h *agentHandle) setupCiliumAgentHive(clientset k8sClient.Clientset, extraC
 			func() *option.DaemonConfig { return option.Config },
 			func() cnicell.CNIConfigManager { return &fakecni.FakeCNIConfigManager{} },
 			func() gc.Enabler { return gc.NewFake() },
+			func() promise.Promise[resource.CRDSync] {
```
Add a e.g. `resource.CRDSyncRejectedPromise` variable and use it here instead of repeating the code in tests.

jrajahalme replied:
Done.

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -1685,6 +1686,7 @@ type daemonParams struct {
 	CGroupManager       cgroup.CGroupManager
 	ServiceResolver     *dial.ServiceResolver
 	Recorder            *recorder.Recorder
+	CrdSyncPromise      promise.Promise[resource.CRDSync]
```
s/Crd/CRD/

jrajahalme replied:
Done.

joamaki commented on pkg/ipam/crd_test.go:
```diff
@@ -59,6 +60,7 @@ func TestIPNotAvailableInPoolError(t *testing.T) {
 }
 
 var testConfigurationCRD = &option.DaemonConfig{
+	ConfigPatchMutex:        new(lock.RWMutex),
```
Makes my eyes bleed that every test now needs to do this. OTOH, maybe this encourages moving away from the global DaemonConfig.. :D

jrajahalme replied:
If we merge https://github.com/cilium/cilium/pull/33164/commits/9a4920592afd704addc333ef46c501b1668f40b7 then most tests would not need this, as then the mutex is only used to protect `Opts`, and it is done in very few places.

I'll look into this in a separate PR.

nebril reviewed (approved):
ci-structure LGTM

chancez reviewed (approved):
Metrics changes LGTM

asauber commented on pkg/option/config.go:
```diff
@@ -3941,16 +3952,18 @@ func (c *DaemonConfig) KubeProxyReplacementFullyEnabled() bool {
-		"agent-runtime-config-1.json",
-		"agent-runtime-config-2.json",
-	}
+	c.shaSum = c.checksum()
```
There is potential for a race here. The hash should be computed after the file is written.

jrajahalme replied:
Moved as you suggested, and also now taking read lock on `c.ConfigPatchMutex` to exclude patching while we store or sum.

asauber replied:
Thanks for the updates. LGTM

asauber commented on pkg/option/config.go:
```diff
@@ -3962,6 +3975,70 @@ func (c *DaemonConfig) StoreInFile(dir string) error {
+		diff = fmt.Errorf("unmarshal failed %q: %w", string(fileBytes), err).Error()
+	} else {
+		// from https://github.com/google/go-cmp/issues/313#issuecomment-1315651560
+		opts := cmp.FilterPath(func(p cmp.Path) bool {
```
Could we include a comment here that the purpose of this filter is to discover the exported struct fields during the comparison?

jrajahalme replied:
Added comment:
```
		// Ignore all unexported fields during Diff.
```

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1737,6 +1739,46 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
+func newSyncCRDsPromise(params syncCRDsPromiseParams) promise.Promise[resource.CRDSync] {
+	crdSyncResolver, crdSyncPromise := promise.New[resource.CRDSync]()
+	if !params.Clientset.IsEnabled() || option.Config.DryMode {
+		crdSyncResolver.Reject(fmt.Errorf("disabled"))
```
Can we provide a more detailed error message, as this would likely end up being logged to the user.

jrajahalme replied:
Now have it like:
```
	crdSyncResolver.Reject(fmt.Errorf("CRDSync promise is disabled in tests"))
```

tommyp1ckles commented on pkg/k8s/resource/crdsync.go:
```diff
@@ -0,0 +1,8 @@
+package resource
+
+// CRDSync is a type used for promise.Promise, passing the error (if any) from the SyncCRDs() call
+// to the result of the promise Await() call.
```
Isn't this what the `Reject(error)` function is for? i.e. What do we want to pass the error via the CRDType (with Resolve(err) sometimes vs Reject)?

jrajahalme replied:
Good question! It would be weird to me is the CRDSync would be an empty type, so decided the pass the errror from `SyncCRDs` as the promise type. 

I'll test if this works with en empty type instead.

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1716,16 +1724,45 @@ func newDaemonPromise(params daemonParams) (promise.Promise[*Daemon], promise.Pr
 			}
-			daemonResolver.Resolve(daemon)
+
+			// 'option.Config' is assumed to be stable at this point, execpt for
```
This isn't necessarily true, startDaemon does performs many routines (most of which related to KPR configuration) that may override vars in `*option.DeamonConfig`.  The intention of this promise was a way to wait for specifically those flags.

[Here's ](https://github.com/cilium/cilium/pull/32558/files#diff-8a383aefc352ba8854dbab293a563fecf7f8916983d013ab36f677293996ca3cR1359-R1397) a list from a draft PR I was working on of all the configs that can change during the execution of `startDaemon`.

tommyp1ckles replied:
Examples of where startDaemon is required is the `finishKubeProxyReplacementInit` which can disable `EnableNodePort` (which currently must be in startDaemon because it waits for device detection to complete).

jrajahalme replied:
It looks to me like `finishKubeProxyReplacementInit()` as well as `disableNodePort()` calls are made in `newDaemon()` and not in `startDaemon()`. So these would not be examples of config changing after the DaemonConfig promise is resolved.

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1989,6 +1989,19 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+		cfgGroup.Name,
+		controller.ControllerParams{
+			Group:  cfgGroup,
+			DoFunc: option.Config.ValidateUnchanged,
```
Can we pass a cell.Health for the `ControllerParams.Health` parameters so we have health reporting for this reconciliation loop?

jrajahalme replied:
Done, thanks!

gandro reviewed (approved):
Approving for recent `IPCache` changes. I have not reviewed the other changes that are unrelated to my codeowners.

giorio94 commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -97,13 +99,53 @@ func registerHooks(lc cell.Lifecycle, params parameters) error {
+	})
+
+	return crdSyncPromise
+}
```
I'd suggest to generalize this logic and extract it into a separate cell which just takes the list of resources to wait for as a parameter. So that we don't duplicate the logic in multiple places, which will inevitably start diverging over time.

jrajahalme replied:
Done, thanks for the suggestion!

giorio94 commented on pkg/k8s/resource_ctors.go:
```diff
@@ -433,24 +450,24 @@ func CiliumExternalWorkloads(lc cell.Lifecycle, cs client.Clientset, opts ...fun
 
-func CiliumEnvoyConfigResource(lc cell.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) (resource.Resource[*cilium_api_v2.CiliumEnvoyConfig], error) {
-	if !cs.IsEnabled() {
+func CiliumEnvoyConfigResource(params CiliumResourceParams, opts ...func(*metav1.ListOptions)) (resource.Resource[*cilium_api_v2.CiliumEnvoyConfig], error) {
```
The `CiliumExternalWorkloads` resource above should be updated as well. And the same also applies to the clustermesh-apiserver specific resource in `clustermesh-apiserver/clustermesh/k8s/resources.go`. More in general, I'm a bit concerned that this is very easy to miss for new resources, but I'm not sure if there's any good solution there.

jrajahalme replied:
Done. I guess having all Cilium resource constructors do this is a good start for later cut&paste!

giorio94 commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -358,7 +401,13 @@ func startServer(
+			err = syncErr
+		}
+		log.WithError(err).Fatal("Wait for CRD resources failed")
+	}
```
This logic could be dropped altogether, as the code below only fetches events using `Resource[T]`.

jrajahalme replied:
Good to know! Done.

giorio94 commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -97,13 +99,53 @@ func registerHooks(lc cell.Lifecycle, params parameters) error {
+			wg.Wait()
+			return nil
+		},
+	})
```
Nit. This could be simplified using a one-shot job.

jrajahalme replied:
Done, thanks for the suggestion!

giorio94 reviewed (changes requested):
I've only reviewed the clustermesh-specific changes, and left a couple of comments inline.

harsimran-pabla reviewed (approved):
BGP change lgtm.

jrajahalme committed:
k8s/synced: Add CRDSync promise

Add CRDSync promise that cells can use to make it safe to request for
Cilium k8s resources.

CRDSync promise is required for the agent and clustermesh-apiserver hives.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

giorio94 commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -359,11 +357,6 @@ func startServer(
-	_, err := crdSyncPromise.Await(startCtx)
-	if err != nil {
-		log.WithError(err).Fatal("Wait for CRD resources failed")
-	}
```
Nit. The changes in this file should be squashed into the previous commit.

giorio94 commented on pkg/k8s/resource/resource.go:
```diff
@@ -354,6 +362,11 @@ func (r *resource[T]) startWhenNeeded() {
 
+	// Wait for CRDs to have synced before trying to access (Cilium) k8s resources
+	if r.opts.crdSyncPromise != nil {
+		r.opts.crdSyncPromise.Await(r.ctx)
```
Nit. Should return in case of an error, as that means that the context expired and the component is shutting down.

jrajahalme replied:
will have to leave for later.

giorio94 reviewed (approved):
clustermesh changes look good to me, thanks! Just a couple of nits inline.
